Fix SIGSEGV in StorageKafka on DROP TABLE

azat opened the pull request:
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix SIGSEGV in StorageKafka on DROP TABLE

Detailed description / Documentation draft:
After #11599 it is possible that messages of the
ReadBufferFromKafkaConsumer will be cleaned-up right in
read_kafka_message callback (from KafkaBlockInputStream) if the stop
flag isset (i.e. DROP TABLE is waiting the consumer), and if
read_kafka_message already processed some rows it will not return 0 and
the loop after will try to get current topic from the buffer, which uses
messages in the underlying and this will got SIGSEGV:

    12:14:56.173262 [ 55421 ] {f7930856-d478-4e41-af56-24ce7b693e95} <Debug> executeQuery: (from 0.0.0.0:0, user: ) DROP TABLE IF EXISTS data.queue
    12:14:56.173285 [ 55421 ] {f7930856-d478-4e41-af56-24ce7b693e95} <Trace> StorageKafka (newly_queue): Waiting for cleanup
    12:14:56.180016 [ 55390 ] {} <Trace> BaseDaemon: Received signal 11
    12:14:56.180267 [ 4914 ] {} <Fatal> BaseDaemon: ########################################
    12:14:56.181879 [ 4914 ] {} <Fatal> BaseDaemon: (version 20.6.1.1, build id: 4CE0298F08583658) (from thread 55468) (no query) Received signal Segmentation fault (11)
    12:14:56.181900 [ 4914 ] {} <Fatal> BaseDaemon: Address: 0x8 Access: read. Address not mapped to object.
    12:14:56.181909 [ 4914 ] {} <Fatal> BaseDaemon: Stack trace:
    12:14:56.184676 [ 4914 ] {} <Fatal> BaseDaemon: 3. /ch/contrib/cppkafka/include/cppkafka/message.h:111: DB::KafkaBlockInputStream::readImpl() @ 0xe343f1c in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.185553 [ 4914 ] {} <Fatal> BaseDaemon: 4. /ch/contrib/libcxx/include/vector:1003: DB::IBlockInputStream::read() @ 0xd9d95bd in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.188238 [ 4914 ] {} <Fatal> BaseDaemon: 5. /ch/src/DataStreams/copyData.cpp:26: DB::copyData() @ 0xd9f712a in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.188780 [ 4914 ] {} <Fatal> BaseDaemon: 6. /ch/contrib/libcxx/include/vector:1532: DB::StorageKafka::streamToViews() @ 0xe335e73 in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.189331 [ 4914 ] {} <Fatal> BaseDaemon: 7. /ch/src/Storages/Kafka/StorageKafka.cpp:491: DB::StorageKafka::threadFunc() @ 0xe336738 in /usr/lib/debug/usr/bin/clickhouse

55421 thread (shows that it still waiting for deactivation):

    5  std::__1::lock_guard<>::lock_guard () at ../contrib/libcxx/include/__mutex_base:90
    6  DB::BackgroundSchedulePoolTaskInfo::deactivate (this=0x7fc7e4465f20) at ../src/Core/BackgroundSchedulePool.cpp:59
    7  DB::StorageKafka::shutdown (this=0x7fc7e45e4600) at ../contrib/libcxx/include/memory:3821

And just in case thread where read_kafka_message is called:

    0  DB::ReadBufferFromKafkaConsumer::nextImpl (this=0x7fd4901d4118) at ../contrib/libcxx/include/atomic:1491
    1  DB::ReadBuffer::next (this=0x7fd4901d4118) at ../src/IO/ReadBuffer.h:59
    2  DB::ReadBuffer::eof (this=0x7fd4901d4118) at ../src/IO/ReadBuffer.h:81
    3  DB::skipWhitespaceIfAny (buf=...) at ../src/IO/ReadHelpers.h:945
    4  DB::JSONEachRowRowInputFormat::readRow (ext=..., columns=..., this=0x7fd499a7a020) at ../src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:222
    5  DB::JSONEachRowRowInputFormat::readRow (this=0x7fd499a7a020, columns=..., ext=...) at ../src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:218
    6  DB::IRowInputFormat::generate (this=0x7fd499a7a020) at ../src/Processors/Formats/IRowInputFormat.cpp:64
    7  DB::ISource::work (this=0x7fd499a7a020) at ../src/Processors/ISource.cpp:48
    8  DB::KafkaBlockInputStream::<lambda()>::operator()(void) const () at ../contrib/libcxx/include/memory:3826
    9  DB::KafkaBlockInputStream::readImpl (this=0x7fd46e718820) at ../contrib/libcxx/include/new:340

Cc: @filimonov

<details>

HEADs:
- 084dabefefdffdfa2125c4bf95bced7275e446e3
- 278061f0b20234787b631397a6fdf1e1243e3180
- db618bbe3fbebebfe979ffa4ffb59e79374524b8

</details>

azat committed:
kafka: fix SIGSEGV on DROP TABLE

After #11599 it is possible that messages of the
ReadBufferFromKafkaConsumer will be cleaned-up right in
read_kafka_message callback (from KafkaBlockInputStream) if the stop
flag isset (i.e. DROP TABLE is waiting the consumer), and if
read_kafka_message already processed some rows it will not return 0 and
the loop after will try to get current topic from the buffer, which uses
messages in the underlying and this will got SIGSEGV:

    12:14:56.173262 [ 55421 ] {f7930856-d478-4e41-af56-24ce7b693e95} <Debug> executeQuery: (from 0.0.0.0:0, user: ) DROP TABLE IF EXISTS data.queue
    12:14:56.173285 [ 55421 ] {f7930856-d478-4e41-af56-24ce7b693e95} <Trace> StorageKafka (newly_queue): Waiting for cleanup
    12:14:56.180016 [ 55390 ] {} <Trace> BaseDaemon: Received signal 11
    12:14:56.180267 [ 4914 ] {} <Fatal> BaseDaemon: ########################################
    12:14:56.181879 [ 4914 ] {} <Fatal> BaseDaemon: (version 20.6.1.1, build id: 4CE0298F08583658) (from thread 55468) (no query) Received signal Segmentation fault (11)
    12:14:56.181900 [ 4914 ] {} <Fatal> BaseDaemon: Address: 0x8 Access: read. Address not mapped to object.
    12:14:56.181909 [ 4914 ] {} <Fatal> BaseDaemon: Stack trace:
    12:14:56.184676 [ 4914 ] {} <Fatal> BaseDaemon: 3. /ch/contrib/cppkafka/include/cppkafka/message.h:111: DB::KafkaBlockInputStream::readImpl() @ 0xe343f1c in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.185553 [ 4914 ] {} <Fatal> BaseDaemon: 4. /ch/contrib/libcxx/include/vector:1003: DB::IBlockInputStream::read() @ 0xd9d95bd in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.188238 [ 4914 ] {} <Fatal> BaseDaemon: 5. /ch/src/DataStreams/copyData.cpp:26: DB::copyData() @ 0xd9f712a in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.188780 [ 4914 ] {} <Fatal> BaseDaemon: 6. /ch/contrib/libcxx/include/vector:1532: DB::StorageKafka::streamToViews() @ 0xe335e73 in /usr/lib/debug/usr/bin/clickhouse
    12:14:56.189331 [ 4914 ] {} <Fatal> BaseDaemon: 7. /ch/src/Storages/Kafka/StorageKafka.cpp:491: DB::StorageKafka::threadFunc() @ 0xe336738 in /usr/lib/debug/usr/bin/clickhouse

55421 thread (shows that it still waiting for deactivation):

    5  std::__1::lock_guard<>::lock_guard () at ../contrib/libcxx/include/__mutex_base:90
    6  DB::BackgroundSchedulePoolTaskInfo::deactivate (this=0x7fc7e4465f20) at ../src/Core/BackgroundSchedulePool.cpp:59
    7  DB::StorageKafka::shutdown (this=0x7fc7e45e4600) at ../contrib/libcxx/include/memory:3821

And just in case thread where read_kafka_message is called:

    0  DB::ReadBufferFromKafkaConsumer::nextImpl (this=0x7fd4901d4118) at ../contrib/libcxx/include/atomic:1491
    1  DB::ReadBuffer::next (this=0x7fd4901d4118) at ../src/IO/ReadBuffer.h:59
    2  DB::ReadBuffer::eof (this=0x7fd4901d4118) at ../src/IO/ReadBuffer.h:81
    3  DB::skipWhitespaceIfAny (buf=...) at ../src/IO/ReadHelpers.h:945
    4  DB::JSONEachRowRowInputFormat::readRow (ext=..., columns=..., this=0x7fd499a7a020) at ../src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:222
    5  DB::JSONEachRowRowInputFormat::readRow (this=0x7fd499a7a020, columns=..., ext=...) at ../src/Processors/Formats/Impl/JSONEachRowRowInputFormat.cpp:218
    6  DB::IRowInputFormat::generate (this=0x7fd499a7a020) at ../src/Processors/Formats/IRowInputFormat.cpp:64
    7  DB::ISource::work (this=0x7fd499a7a020) at ../src/Processors/ISource.cpp:48
    8  DB::KafkaBlockInputStream::<lambda()>::operator()(void) const () at ../contrib/libcxx/include/memory:3826
    9  DB::KafkaBlockInputStream::readImpl (this=0x7fd46e718820) at ../contrib/libcxx/include/new:340

Cc: @filimonov

filimonov commented:
Ok, i think i understand what happened. I think it was not between poll and `read_kafka_message` (in that case next() call would just return false, 0 rows world be parsed and it would never enter `if (new_rows)`, 
but stop should happened inside `read_kafka_message` itself. First call of `next()` function return the message and it's parsed, after that it call `next()` one more time (to ensure that it's on eof), and in that moment `next()` clean up messages inside `resetIfStopped`. And after that we have a situation then some rows were parsed, but message was already destructed. 

In other words - i think that your fix will not help. I would try smth like that:

```diff
diff --git a/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp b/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
index 3fd28cde5e..fc239d422d 100644
--- a/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
+++ b/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
@@ -424,9 +424,13 @@ bool ReadBufferFromKafkaConsumer::nextImpl()
     /// NOTE: ReadBuffer was implemented with an immutable underlying contents in mind.
     ///       If we failed to poll any message once - don't try again.
     ///       Otherwise, the |poll_timeout| expectations get flawn.
+
+    if (!allowed)
+        return false;
+
     resetIfStopped();
 
-    if (!allowed || !hasMorePolledMessages())
+    if (!hasMorePolledMessages())
         return false;
 
     // XXX: very fishy place with const casting.
```

azat commented:
>Ok, i think i understand what happened. I think it was not between poll and read_kafka_message (in that case next() call would just return false, 0 rows world be parsed and it would never enter if (new_rows),
but stop should happened inside read_kafka_message itself. First call of next() function return the message and it's parsed, after that it call next() one more time (to ensure that it's on eof), and in that moment next() clean up messages inside resetIfStopped. And after that we have a situation then some rows were parsed, but message was already destructed.

Yes
> "messages of the ReadBufferFromKafkaConsumer will be cleaned-up right in read_kafka_message callback (from KafkaBlockInputStream) if the stop flag isset"

>i think that your fix will not help

But you are right it is still possible to get empty messages

> I would try smth like that:

Yes, this should work, but looks messy (too much flags in that reader), how about not checking stop flag in the `ReadBufferFromKafkaConsumer::nextImpl`?

(NOTE: that `consumerStopped()` is still required, since otherwise it will retry poll)

filimonov commented:
> Yes, this should work, but looks messy (too much flags in that reader), how about not checking stop flag in the ReadBufferFromKafkaConsumer::nextImpl?

+1. Lets do it. Had the same thoughts. 

> (NOTE: that consumerStopped() is still required, since otherwise it will retry poll)

It is not. Check that line:
https://github.com/azat/ClickHouse/blob/278061f0b20234787b631397a6fdf1e1243e3180/src/Storages/Kafka/KafkaBlockInputStream.cpp#L202

azat commented:
>It is not. Check that line:

It will go to the first branch https://github.com/azat/ClickHouse/blob/278061f0b20234787b631397a6fdf1e1243e3180/src/Storages/Kafka/KafkaBlockInputStream.cpp#L198 (remember `stalled_status == CONSUMER_STOPPED`, so `buffer->isStalled() == true`), although not sure this is ok in itself

filimonov commented:
Strange, and seems it has no sence. (last if branch in dead in that case).

The intent was different, i can recheck that tomorrow or you can just try to switch the order of if else to check dataunusable first.

filimonov commented:
Yes, it seems like logical error (the behavior of functions isStalled changed during refactorings, and that part of code was not fixed accordingly). 

I've pushed a commit to your branch see https://github.com/ClickHouse/ClickHouse/pull/12075/commits/821317e6273ea3e9f790bec4162ac27986c69571

azat commented:
>Yes, it seems like logical error

Ok, squashed

filimonov reviewed (approved):
LGFM

azat commented:
Should be backported to 20.5 (#11870 )

akuzm commented on src/Storages/Kafka/KafkaBlockInputStream.cpp:
```diff
@@ -186,14 +186,14 @@ Block KafkaBlockInputStream::readImpl()
+        else if (buffer->isStalled())
+        {
+            ++failed_poll_attempts;
+        }
```
So, reordering these blocks fixes the segfault? Is it possible to add an explanation why this order matters and why it must be this way and not some another way?

azat replied:
Not really, the SIGSEGV is fixed by avoiding cleaning messages while parsing (i.e. removing resetIfStopped from ReadBufferFromKafkaConsumer::nextImpl).

This reordering is to avoid extra polling if stop flag isset, i.e. prefer break over incrementing failed poll attempts

akuzm replied:
Maybe we can add some comment there, then? The reason why I'm asking is that this kind of fix is always suspicious: one line removed, segfault is fixed, but there was no comment justifying why it was there, and there is no comment justifying why it must not be there. It made sense to add this line before, maybe we should write down why it was a bad idea, for the future.

Of course, it might be difficult to write a comment for line that does not exist, so if a useful comment doesn't come to mind, let's just leave it as is.

Can we add an integration test for this?

filimonov replied:
Here is the fix for segfault and explanation: https://github.com/ClickHouse/ClickHouse/pull/12075#issuecomment-652630127

Reproduce in the integration test reliable is almost impossible (DROP TABLE should happen in a very specific moment), there are quite a lot of tests doing table drops during consumption, they did not catch the bug. 

It may be possible in the unit test, but it would require mocking a lot of stuff, and the required effort is too big. 

Moving condition lower - is was just a logical error which was leading to a few unneeded retries instead of leaving the loop immediately. (unrelated to the segfault).

azat replied:
>Maybe we can add some comment there, then? The reason why I'm asking is that this kind of fix is always suspicious: one line removed, segfault is fixed, but there was no comment justifying why it was there, and there is no comment justifying why it must not be there. It made sense to add this line before

@akuzm Agree, such fixes are always very suspicious.
Comments does not always helps to avoid with re-introducing bugs, since something similar can be added to some helper and/or not always comments will be read.
I think that `assert`/`LOGICAL_ERROR` will be a better idea, will add

>Can we add an integration test for this?

Like @filimonov told, it will be racy

azat replied:
Rebased and splitted into separate patches

filimonov commented on src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:
```diff
@@ -420,12 +420,9 @@ void ReadBufferFromKafkaConsumer::resetIfStopped()
-
     /// NOTE: ReadBuffer was implemented with an immutable underlying contents in mind.
     ///       If we failed to poll any message once - don't try again.
     ///       Otherwise, the |poll_timeout| expectations get flawn.
```
That comment old and unneeded.

azat committed:
kafka: improve logging during engine shutdown

This will help with tracking possible issues, when you need to know was
buffer released or not.

azat committed:
kafka: avoid superior polling after DROP/DETACH TABLE

Before this patch isStalled() was checked before polledDataUnusable(),
and after DROP TABLE isStalled() == true (although this looks tricky).

azat committed:
kafka: check that the data is still usable after parsing

azat committed:
kafka: remove outdated comment

As stated by @filimonov it is not relevant (after #11599)

filimonov commented on src/Storages/Kafka/KafkaBlockInputStream.cpp:
```diff
@@ -186,17 +193,17 @@ Block KafkaBlockInputStream::readImpl()
+        }
         else
         {
-            LOG_WARNING(log, "Parsing of message (topic: {}, partition: {}, offset: {}) return no rows.", buffer->currentTopic(), buffer->currentPartition(), buffer->currentOffset());
```
what happened with that line?

azat replied:
Because it may also bail, i.e. what will happen if buffer will be in REBALANCE_HAPPENED state?

filimonov replied:
Then if branch above that line will catch it.

filimonov replied:
If think you're afraid to get that state because of rebalance callback. But it is synchronious and called only during poll

azat replied:
>Then if branch above that line will catch it.

This states are too tricky, but you are right

filimonov commented on src/Storages/Kafka/KafkaBlockInputStream.cpp:
```diff
@@ -140,6 +140,13 @@ Block KafkaBlockInputStream::readImpl()
+            // will be called, that may make something unusable, i.e. clean
+            // ReadBufferFromKafkaConsumer::messages, which is accessed from
+            // ReadBufferFromKafkaConsumer::currentTopic() (and other helpers).
+            if (buffer->polledDataUnusable())
```
actually if it's stalled you can access currentTopic etc.

azat replied:
How? Let's say that buffer is in REBALANCE_HAPPENED state, so no messages, and it will try to access them via iterator to get the topic

filimonov replied:
Then if branch above that line will catch it.

filimonov replied:
> actually if it's stalled you can access currentTopic etc.

actually if it's stalled you >>>can't<<< access currentTopic etc.

Silly error. I wanted to say "check isStalled indead of polledDataUnusable"
